grails-datastore-core/src/main/groovy/org/grails/datastore/mapping/services/SoftServiceLoader.java

```diff
@@ -0,0 +1,268 @@
+/*
+ * Copyright 2017-2020 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.grails.datastore.mapping.services;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.slf4j.helpers.NOPLogger;
+import org.springframework.util.ClassUtils;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.URL;
+import java.util.*;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+/**
+ * <p>Variation of {@link java.util.ServiceLoader} that allows soft loading and conditional loading of
+ * META-INF/services classes.</p>
+ *
+ * @param <S> The service type
+ * @author Graeme Rocher
+ * @since 1.0
+ */
+public final class SoftServiceLoader<S> implements Iterable<ServiceDefinition<S>> {
```

Comment:
Was this class in each of the implementations? How was this being done before?

Reply:
Yes, this is used in all implementations. Earlier this was done in the `DefaultServiceRegistry` constructor to load the services using `ServiceLoader.loadServices` which is then used in `Datastore::getServices()` method.